File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.kt
package org.example

import java.util.StringTokenizer

fun main() {
    val br = System.`in`.bufferedReader()
    // 루트 없는 트리 루트 1로 가정, 각 노드의 부모
    val N = br.readLine().toInt()
    val graph = Array(N + 1) { mutableListOf<Int>() }
    repeat(N - 1) {
        val st = StringTokenizer(br.readLine())
        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()
        graph[u].add(v)
        graph[v].add(u)
    }

    val visited = IntArray(N + 1)
    visited[1] = -1 //루트 표시
    dfs(graph, 1, visited)

    val sb = StringBuilder()
    for (i in 2..N) {
        sb.append(visited[i]).append("\n")
    }
    print(sb)
    br.close()
}

fun dfs(graph: Array<MutableList<Int>>, v: Int, visited: IntArray) {
    for (neighbor in graph[v]) {
        if (visited[neighbor] == 0) {
            // 방문 전이면
            visited[neighbor] = v // 부모 넣어주고 다음 진행
            dfs(graph, neighbor, visited)
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline


def dfs(graph, v, visited, parent):
    visited[v] = True
    for neighbor in graph[v]:
        if not visited[neighbor]:
            parent[neighbor] = v  # 탐색할 neighbor의 parent는 현재 v가 됨
            dfs(graph, neighbor, visited, parent)


n = int(input())
graph = [[] for _ in range(n + 1)]  # 0은 사용x
parent = [0] * (n + 1)
visited = [False] * (n + 1)
for _ in range(n - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for v in range(1, n + 1):
    if not visited[v]:
        dfs(graph, v, visited, parent)

print('\n'.join(map(str, parent[2:])))
